### PR TITLE
Add Cellsnp to modules

### PR DESCRIPTION
**Pull request by wxicu**
## PR checklist

Related to #5072 
Add cellsnp-lite.
- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [x] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/cellsnp/main.nf [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    tuple val(meta), path(bam), path(bai), path(region_vcf), path(barcode), path(sample_list)
+
+    output:
+    tuple val(meta), path('*/cellSNP.base.vcf.gz')     , emit: base
+    tuple val(meta), path('*/cellSNP.cells.vcf.gz')    , emit: cell,          optional: true
+    tuple val(meta), path('*/cellSNP.samples.tsv')     , emit: sample
+    tuple val(meta), path('*/cellSNP.tag.AD.mtx')      , emit: allele_depth
+    tuple val(meta), path('*/cellSNP.tag.DP.mtx')      , emit: depth_coverage
+    tuple val(meta), path('*/cellSNP.tag.OTH.mtx')     , emit: depth_other
+    path 'versions.yml'                                , emit: versions
```
Comment (tstoeriko) [inline modules/nf-core/cellsnp/main.nf:20]:
```suggestion
    tuple val(meta), path('*/cellSNP.base.vcf.gz') , emit: base
    tuple val(meta), path('*/cellSNP.cells.vcf.gz'), emit: cell          , optional: true
    tuple val(meta), path('*/cellSNP.samples.tsv') , emit: sample
    tuple val(meta), path('*/cellSNP.tag.AD.mtx')  , emit: allele_depth
    tuple val(meta), path('*/cellSNP.tag.DP.mtx')  , emit: depth_coverage
    tuple val(meta), path('*/cellSNP.tag.OTH.mtx') , emit: depth_other
    path 'versions.yml'                            , emit: versions
```

**Review by tstoeriko — COMMENTED** [review]
Good job so far! I've made some suggestions, but I'm still new to nf-core, so it's probably a good idea to have someone else look at the module, too.

**Comment by wxicu** [comment]
@tstoeriko thanks for your detailed suggestion! I have now splitted cellsnp into two submodules for mode a and b (also according to the documentation[https://cellsnp-lite.readthedocs.io/en/latest/main/manual.html]) to control the input type. Both Mode 1a and mode 2a requie only a single bam file. hope this will make the structure more clear.
I have also tried to make the output files included in the snapshot as many as possible

File: modules/nf-core/cellsnp/modea/main.nf [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    tuple val(meta), path(bam), path(bai), path(region_vcf), path(barcode)
+
+    output:
+    tuple val(meta), path('*.base.vcf.gz') , emit: base
+    tuple val(meta), path('*.cells.vcf.gz'), emit: cell          , optional: true
+    tuple val(meta), path('*.samples.tsv') , emit: sample
+    tuple val(meta), path('*.tag.AD.mtx')  , emit: allele_depth
+    tuple val(meta), path('*.tag.DP.mtx')  , emit: depth_coverage
+    tuple val(meta), path('*.tag.OTH.mtx') , emit: depth_other
+    path 'versions.yml'                            , emit: versions
```
Comment (SPPearce) [inline modules/nf-core/cellsnp/modea/main.nf:20]: Can you align the ,emit please.

File: modules/nf-core/cellsnp/modea/main.nf [hunk 3]
```diff
@@ -0,0 +1,68 @@
+process CELLSNP_MODEA {
+    tag "$meta.id"
+    label 'process_high'
```
Comment (SPPearce) [inline modules/nf-core/cellsnp/modea/main.nf:3]: Does this need process high?

File: modules/nf-core/cellsnp/modea/meta.yml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      type: file
+      description: A single BAM/SAM/CRAM file, e.g., from CellRanger.
+      pattern: "*.{bam,cram,sam}"
+  - bai:
+      type: file
+      description: The index of the BAM/CRAM file.
+      pattern: "*.{bai,crai}"
+  - region_vcf:
+      type: file
+      description: A optional vcf file listing all candidate SNPs, for fetch each variants.
```
Comment (SPPearce) [inline modules/nf-core/cellsnp/modea/meta.yml:35]: What does: "for fetch each variants" mean?

File: modules/nf-core/cellsnp/modea/tests/main.nf.test [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        then {
+            assertAll(
+            { assert process.success },
+            { assert snapshot(
+                    process.out.sample,
+                    process.out.allele_depth,
+                    process.out.depth_coverage,
+                    process.out.depth_other).match("stub")
+            },
+	        { assert path(process.out.base.get(0).get(1)).exists() } )
```
Comment (SPPearce) [inline modules/nf-core/cellsnp/modea/tests/main.nf.test:85]: Stub test can just check `process.out`, as nothing should vary.

File: modules/nf-core/cellsnp/modea/tests/main.nf.test [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        when {
+            process {
+                """ 
+                input[0] = SAMTOOLS_INDEX.out.bai.collect{ meta, bai -> bai }.map{ 
+                            bai -> [[ id: 'sample1'], 
+                                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/chr21.bam', checkIfExists: true),
+                                    bai,
+                                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/donor_genotype_chr21.vcf', checkIfExists: true),
+                                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/barcodes.tsv', checkIfExists: true)                                    ]}
```
Comment (SPPearce) [inline modules/nf-core/cellsnp/modea/tests/main.nf.test:36]: I think this should be (appropriately formatted):

```
[[ id: 'sample1'], 
file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/chr21.bam', checkIfExists: true),
SAMTOOLS_INDEX.out.bai.map{it[1]},
file(params.modules_testdata_base_path + /genomics/homo_sapiens/demultiplexing/donor_genotype_chr21.vcf', checkIfExists: true),
file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/barcodes.tsv', checkIfExists: true)]
```

**Review by SPPearce — COMMENTED** [review]
This looks very nice.
I have a question though, [the docs talk about 4 modes](https://cellsnp-lite.readthedocs.io/en/latest/main/manual.html#quick-usage-for-single-cell-or-bulk-rna-seq-data), calling them Mode1a, Mode1b, Mode2a, Mode2b. Which is this? Should it be e.g. mode1a rather than modea?

**Comment by SPPearce** [comment]
> This looks very nice. I have a question though, [the docs talk about 4 modes](https://cellsnp-lite.readthedocs.io/en/latest/main/manual.html#quick-usage-for-single-cell-or-bulk-rna-seq-data), calling them Mode1a, Mode1b, Mode2a, Mode2b. Which is this? Should it be e.g. mode1a rather than modea?

Oh, I see that you have indeed answered this in the comment above, so this is appropriate for either mode1a or mode2a.
Just a few minor things then.

File: modules/nf-core/cellsnp/modea/tests/main.nf.test [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        when {
+            process {
+                """ 
+                
+                input[0] = [[ id: 'sample1'], 
+                            file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/chr21.bam', checkIfExists: true),
+                            SAMTOOLS_INDEX.out.bai.map{it[1]},
+                            file(params.modules_testdata_base_path + /genomics/homo_sapiens/demultiplexing/donor_genotype_chr21.vcf', checkIfExists: true),
```
Comment (SPPearce) [inline modules/nf-core/cellsnp/modea/tests/main.nf.test:35]:
```suggestion
                            file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/donor_genotype_chr21.vcf', checkIfExists: true),
```

File: modules/nf-core/cellsnp/modea/tests/main.nf.test (hunk 5 above)
Comment (SPPearce) [reply modules/nf-core/cellsnp/modea/tests/main.nf.test:85]: Oh, true. But you should be able to get around that by swapping to  `echo "" | gzip > ${prefix}.base.vcf.gz`

File: modules/nf-core/cellsnp/modea/tests/main.nf.test (hunk 7 above)
Comment (wxicu) [reply modules/nf-core/cellsnp/modea/tests/main.nf.test]: @SPPearce I tried your suggestion but didnt pass the tests due to the error ` Not a valid path value type: groovyx.gpars.dataflow.DataflowVariable (DataflowVariable(value=null))` . It seems like I was trying to add a channel to the input channel. So i go back to the previous version now

**Review by SPPearce — APPROVED** [review]

**Pull request merged**